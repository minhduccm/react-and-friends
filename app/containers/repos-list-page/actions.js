/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_REPOS_LIST,
  LOAD_REPOS_LIST_SUCCESS,
  LOAD_REPOS_LIST_ERROR,
  SELECT_REPO,
} from './constants';

/**
 * Load the repositories, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_REPOS
 */
export function loadReposList() {
  return {
    type: LOAD_REPOS_LIST,
  };
}

/**
 * Dispatched when the repositories are loaded by the request saga
 *
 * @param  {array} repos The repository data
 * @param  {string} username The current username
 *
 * @return {object}      An action object with a type of LOAD_REPOS_SUCCESS passing the repos
 */
export function reposListLoaded(repos, username) {
  return {
    type: LOAD_REPOS_LIST_SUCCESS,
    repos,
    username,
  };
}

/**
 * Dispatched when loading the repositories fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_REPOS_ERROR passing the error
 */
export function reposListLoadingError(error) {
  return {
    type: LOAD_REPOS_LIST_ERROR,
    error,
  };
}

/**
 * Select the repository
 *
 * @return {object} An action object with a type of SELECT_REPO
 */
export function selectRepo(selectedRepo) {
  console.log('selectedRepo action');
  return {
    type: SELECT_REPO,
    selectedRepo,
  };
}

