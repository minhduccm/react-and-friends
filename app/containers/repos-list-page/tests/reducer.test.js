/**
 * Repos list page reducer
 */

import reposListReducer from '../reducer';
import { loadReposList, reposListLoaded, reposListLoadingError, selectRepo } from '../actions';

describe('reposListReducer', () => {
  const initialState = reposListReducer(undefined, {});

  it('returns the initial state', () => {
    expect(initialState).toMatchSnapshot();
  });

  it('handles loadReposList action', () => {
    expect(reposListReducer(initialState, loadReposList()))
      .toMatchSnapshot();
  });

  it('handles reposListLoaded action', () => {
    const repos = [{
      id: '999xxx',
      name: 'react-and-friends',
    }, {
      id: '888yyy',
      name: 'flashcard',
    }];
    const username = 'minhduccm';
    expect(reposListReducer(initialState, reposListLoaded(repos, username)))
      .toMatchSnapshot();
  });

  it('handles reposListLoadingError action', () => {
    const error = 'Something went wrong!!!';
    expect(reposListReducer(initialState, reposListLoadingError(error)))
      .toMatchSnapshot();
  });

  it('handles selectRepo action', () => {
    const selectedRepo = {
      id: '777xyz',
      name: 'awesome react app',
    };
    expect(reposListReducer(initialState, selectRepo(selectedRepo)))
      .toMatchSnapshot();
  });
});
