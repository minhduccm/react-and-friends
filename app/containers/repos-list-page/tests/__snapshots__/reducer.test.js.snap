exports[`reposListReducer handles loadReposList action 1`] = `
Object {
  "currentUser": false,
  "error": false,
  "loading": true,
  "selectedRepo": false,
  "userData": Object {
    "repositories": false,
  },
}
`;

exports[`reposListReducer handles reposListLoaded action 1`] = `
Object {
  "currentUser": "minhduccm",
  "error": false,
  "loading": false,
  "selectedRepo": false,
  "userData": Object {
    "repositories": Array [
      Object {
        "id": "999xxx",
        "name": "react-and-friends",
      },
      Object {
        "id": "888yyy",
        "name": "flashcard",
      },
    ],
  },
}
`;

exports[`reposListReducer handles reposListLoadingError action 1`] = `
Object {
  "currentUser": false,
  "error": "Something went wrong!!!",
  "loading": false,
  "selectedRepo": false,
  "userData": Object {
    "repositories": false,
  },
}
`;

exports[`reposListReducer handles selectRepo action 1`] = `
Object {
  "currentUser": false,
  "error": false,
  "loading": false,
  "selectedRepo": Object {
    "id": "777xyz",
    "name": "awesome react app",
  },
  "userData": Object {
    "repositories": false,
  },
}
`;

exports[`reposListReducer returns the initial state 1`] = `
Object {
  "currentUser": false,
  "error": false,
  "loading": false,
  "selectedRepo": false,
  "userData": Object {
    "repositories": false,
  },
}
`;
